### ISSUE TITLE
SPTrustedIdentityTokenIssuer functionality for add/update/remove ProviderRealms

### ISSUE THREAD

**Issue by lazda** [PowerShell/SharePointDsc#593]
Please add functionality to add/update/remove ProviderRealms for SPTrustedIdentityTokenIssuer resource.

Version of the Operating System and PowerShell the DSC Target Node is running:
WinSrv 2016

Version of the DSC module you're using:
1.6

**Comment by lazda** [comment]
Example:
```
$ap = Get-SPTrustedIdentityTokenIssuer
$uri = new-object System.Uri("https://<appname>.domain.local")
$ap.ProviderRealms.Add($uri, "urn:sharepoint:<appname>")
$ap.Update()
```

**Comment by lazda** [comment]
Here is my solution:
MSFT_SPProviderRealms.psm1:
```
function Get-TargetResource
{
    [CmdletBinding()]
    [OutputType([System.Collections.Hashtable])]
    param
    (
        [parameter(Mandatory = $true)]
        [string]
        $IssuerName,
		
		[parameter(Mandatory = $true)]
        [Microsoft.Management.Infrastructure.CimInstance[]]
        $ProviderRealms,

        [parameter(Mandatory = $false)]
        [ValidateSet("Present","Absent")]
        [String]
        $Ensure = "Present",

        [parameter(Mandatory = $false)]
        [System.Management.Automation.PSCredential]
        $InstallAccount
    )

    Write-Verbose -Message "Getting SPTrustedIdentityTokenIssuer ProviderRealms"
	
    $result = Invoke-SPDSCCommand -Credential $InstallAccount `
                                  -Arguments $PSBoundParameters `
                                  -ScriptBlock {
        $params = $args[0]
		
        $providerRealms = New-Object 'system.collections.generic.dictionary[system.uri,string]'
		
		foreach($key in $params.ProviderRealms)
		{
			
			$url=New-Object System.Uri($key.Key)

			if(!!$url.AbsoluteUri -and "$($key.Value)".ToLower().StartsWith("urn:"))
			{
				$providerRealms[$url] = $key.Value
			}
			else
			{
				throw "Can't convert '$($key.Key)' to Uri or Urn value is worng. Please check your configuration file"
			}
		}
		
		
        $spTrust = Get-SPTrustedIdentityTokenIssuer -Identity $params.IssuerName `
                                                    -ErrorAction SilentlyContinue
		
		if (!$spTrust)
		{
			throw "SPTrustedIdentityTokenIssuer '$($params.IssuerName)' not found"
		}
		
		$diffObjects = [Linq.Enumerable]::Intersect($providerRealms, $($spTrust.ProviderRealms))
		
		if($params.Ensure -eq "Present")
		{
			$present = $($diffObjects).Count -eq $($providerRealms).Count
		}
		else
		{
			$present = !$($($diffObjects).Count -eq 0)
		}
		
		$currentState = @{$true = "Present"; $false = "Absent"}[$present]
		
        return @{
            IssuerName                   = $params.IssuerName
            ProviderRealms               = $spTrust.ProviderRealms
            Ensure                       = $currentState
        }        
    }
    return $result
}

function Set-TargetResource
{
    [CmdletBinding()]
    param
    (
        [parameter(Mandatory = $true)]
        [string]
        $IssuerName,
		
		[parameter(Mandatory = $true)]
        [Microsoft.Management.Infrastructure.CimInstance[]]
        $ProviderRealms,

        [parameter(Mandatory = $false)]
        [ValidateSet("Present","Absent")]
        [String]
        $Ensure = "Present",

        [parameter(Mandatory = $false)]
        [System.Management.Automation.PSCredential]
        $InstallAccount
    )
	
    $CurrentValues = Get-TargetResource @PSBoundParameters

    if ($Ensure -eq "Present") 
    {
        if ($CurrentValues.Ensure -eq "Absent")
        {
			

            Write-Verbose -Message "Setting SPTrustedIdentityTokenIssuer provider realms"

            $result = Invoke-SPDSCCommand -Credential $InstallAccount `
                                          -Arguments $PSBoundParameters `
                                          -ScriptBlock {
                $params = $args[0]
                
				$trust = Get-SPTrustedIdentityTokenIssuer -Identity $params.IssuerName `
                                                    -ErrorAction SilentlyContinue

                foreach($cKey in $params.ProviderRealms)
				{

					$url= [System.Uri]$cKey.Key
					if ($trust.ProviderRealms.ContainsKey($url))
					{
						if($trust.ProviderRealms[$url] -ne $cKey.Value)
						{
							Write-Verbose -Message "The provider realm '$($cKey.Key)' exists but has different value. Replacing."
							$trust.ProviderRealms.Remove($url)						
							$trust.ProviderRealms.Add($url, $cKey.Value)
						}
						else
						{
							Write-Verbose -Message "Provider realm '$($cKey.Key)' exists. Skipping."
						}
					}
					else
					{
						Write-Verbose -Message "Adding new provider realm '$($cKey.Key)'"
						$trust.ProviderRealms.Add($url, $cKey.Value)
						
					}
				}

                $trust.Update()
            }
        }
    }
    else
    {
        Write-Verbose "Removing SPTrustedIdentityTokenIssuer provider realms"
        $result = Invoke-SPDSCCommand -Credential $InstallAccount `
                                      -Arguments $PSBoundParameters `
                                      -ScriptBlock {
            $params = $args[0]
			$update = $false
            $trust = Get-SPTrustedIdentityTokenIssuer -Identity $params.IssuerName `
                                                    -ErrorAction SilentlyContinue

                foreach($cKey in $params.ProviderRealms)
				{
					$url=[System.Uri]$cKey.Key
					
					if ($trust.ProviderRealms.ContainsKey($url))
					{
						Write-Verbose -Message "Removing provider realm '$($cKey.Key)'."
						$trust.ProviderRealms.Remove($url)
						$update = $true
					}
				}
				
				if($update)
				{
					$trust.Update()
				}
				else
				{
					throw "Provider realm '$($cKey.Key)' does not exist."
				}
        }
    }
}

function Test-TargetResource
{
    [CmdletBinding()]
    [OutputType([Boolean])]
    param
    (
        [parameter(Mandatory = $true)]
        [string]
        $IssuerName,
		
		[parameter(Mandatory = $true)]
        [Microsoft.Management.Infrastructure.CimInstance[]]
        $ProviderRealms,

        [parameter(Mandatory = $false)]
        [ValidateSet("Present","Absent")]
        [String]
        $Ensure = "Present",

        [parameter(Mandatory = $false)]
        [System.Management.Automation.PSCredential]
        $InstallAccount
    )

    Write-Verbose -Message "Testing SPTrustedIdentityTokenIssuer provider realms"
	
    $CurrentValues = Get-TargetResource @PSBoundParameters
	
    return Test-SPDscParameterState -CurrentValues $CurrentValues `
                                    -DesiredValues $PSBoundParameters `
                                    -ValuesToCheck @("Ensure")
}

Export-ModuleMember -Function *-TargetResource
```
MSFT_SPProviderRealms.schema.mof:

```
[ClassVersion("1.1.0.0"), FriendlyName("SPProviderRealms")]
class MSFT_SPProviderRealms : OMI_BaseResource
{
    [Key, Description("Name of the SPTrustedIdentityTokenIssuer")] String IssuerName;
    [Required, EmbeddedInstance("MSFT_KeyValuePair"), Description("Hashtable with Provider Realms that is passed to identity provider")] String ProviderRealms[];
    [Write, Description("Present if the ProviderRealms should be created, or Absent if it should be removed"), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Write, Description("POWERSHELL 4 ONLY: The account to run this resource as, use PsDscRunAsCredential if using PowerShell 5"), EmbeddedInstance("MSFT_Credential")] String InstallAccount;
};
```

**Comment by ykuijs** [comment]
Closed via PR #754
